src/swarm-topography/Swarm/Game/Scenario/Topography/ProtoCell.hs

```diff
@@ -0,0 +1,41 @@
+{-# LANGUAGE DerivingVia #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+-- |
+-- SPDX-License-Identifier: BSD-3-Clause
+module Swarm.Game.Scenario.Topography.ProtoCell (
+  SignpostableCell (..),
+  StructurePalette (..),
+) where
+
+import Control.Applicative ((<|>))
+import Data.Aeson.KeyMap (KeyMap)
+import Data.Yaml as Y
+import Swarm.Game.Scenario.Topography.Navigation.Waypoint (WaypointConfig)
+import Swarm.Util.Yaml
+
+newtype StructurePalette e = StructurePalette
+  {unPalette :: KeyMap (SignpostableCell e)}
+  deriving (Eq, Show)
+
+instance (FromJSONE e a) => FromJSONE e (StructurePalette a) where
+  parseJSONE =
+    withObjectE "palette" $
+      fmap StructurePalette . mapM parseJSONE
+
+-- | Supplements a cell with waypoint information
+data SignpostableCell c = SignpostableCell
+  { waypointCfg :: Maybe WaypointConfig
+  , standardCell :: c
+  }
+  deriving (Eq, Show)
+
+instance (FromJSONE e a) => FromJSONE e (SignpostableCell a) where
+  parseJSONE x =
+    withObjectE "SignpostableCell" objParse x
+      <|> (SignpostableCell Nothing <$> parseJSONE x)
+   where
+    objParse v =
+      SignpostableCell
+        <$> liftE (v .:? "waypoint")
```

Comment:
After #1881 I think we are going to prefer the style using `do`-notation with named fields and `RecordWildcards` to assemble the final result, like so (untested):
```
objParse v = do
  waypointCfg <- liftE (v .:? "waypoint")
  standardCell <- v ..: "cell"
  pure SignpostableCell {..}

Reply:
Agreed, I like the new style, and used it [here](https://github.com/swarm-game/swarm/pull/1924/files#r1641329560).  I'll fix this one, too.